File: protocol/src/test/go/zfoogo/NormalObject.go
package protocol

type NormalObject struct {
	A int8
    Aaa []int8
    B int16
    C int
    D int64
    E float32
    F float64
    G bool
    Jj string
    Kk ObjectA
    L []int
    Ll []int64
    Lll []ObjectA
    Llll []string
    M map[int]string
    Mm map[int]ObjectA
    S []int
    Ssss []string
    OutCompatibleValue int
}

func (protocol NormalObject) ProtocolId() int16 {
	return 101
}

func (protocol NormalObject) write(buffer *ByteBuffer, packet any) {
	if packet == nil {
	    buffer.WriteInt(0)
		return
	}
	var message = packet.(*NormalObject)
	var beforeWriteIndex = buffer.WriteOffset()
    buffer.WriteInt(854)
    buffer.WriteByte(message.A)
    buffer.WriteByteArray(message.Aaa)
    buffer.WriteShort(message.B)
    buffer.WriteInt(message.C)
    buffer.WriteLong(message.D)
    buffer.WriteFloat(message.E)
    buffer.WriteDouble(message.F)
    buffer.WriteBool(message.G)
    buffer.WriteString(message.Jj)
    buffer.WritePacket(&message.Kk, 102)
    buffer.WriteIntArray(message.L)
    buffer.WriteLongArray(message.Ll)
    if message.Lll == nil {
        buffer.WriteInt(0)
    } else {
        buffer.WriteInt(len(message.Lll))
        var length0 = len(message.Lll)
        for i1 := 0; i1 < length0; i1++ {
            var element2 = message.Lll[i1]
            buffer.WritePacket(&element2, 102)
        }
    }
    buffer.WriteStringArray(message.Llll)
    buffer.WriteIntStringMap(message.M)
    if (message.Mm == nil) || (len(message.Mm) == 0) {
        buffer.WriteInt(0)
    } else {
        buffer.WriteInt(len(message.Mm))
        for keyElement4, valueElement5 := range message.Mm {
            buffer.WriteInt(keyElement4)
            buffer.WritePacket(&valueElement5, 102)
        }
    }
    buffer.WriteIntArray(message.S)
    buffer.WriteStringArray(message.Ssss)
    buffer.WriteInt(message.OutCompatibleValue)
    buffer.AdjustPadding(854, beforeWriteIndex)
}

func (protocol NormalObject) read(buffer *ByteBuffer) any {
	var packet = new(NormalObject)
	var length = buffer.ReadInt()
	if length == 0 {
		return packet
	}
	var beforeReadIndex = buffer.ReadOffset()
	var result6 = buffer.ReadByte()
    packet.A = result6
    var array7 = buffer.ReadByteArray()
    packet.Aaa = array7
    var result8 = buffer.ReadShort()
    packet.B = result8
    var result9 = buffer.ReadInt()
    packet.C = result9
    var result10 = buffer.ReadLong()
    packet.D = result10
    var result11 = buffer.ReadFloat()
    packet.E = result11
    var result12 = buffer.ReadDouble()
    packet.F = result12
    var result13 = buffer.ReadBool()
    packet.G = result13
    var result14 = buffer.ReadString()
    packet.Jj = result14
    var result15 = *buffer.ReadPacket(102).(*ObjectA)
    packet.Kk = result15
    var list16 = buffer.ReadIntArray()
    packet.L = list16
    var list17 = buffer.ReadLongArray()
    packet.Ll = list17
    var size20 = buffer.ReadInt()
    var result18 = make([]ObjectA, size20)
    if size20 > 0 {
        for index19 := 0; index19 < size20; index19++ {
            var result21 = *buffer.ReadPacket(102).(*ObjectA)
            result18[index19] = result21
        }
    }
    packet.Lll = result18
    var list22 = buffer.ReadStringArray()
    packet.Llll = list22
    var map23 = buffer.ReadIntStringMap()
    packet.M = map23
    var size25 = buffer.ReadInt()
    var result24 = make(map[int]ObjectA)
    if size25 > 0 {
        for index26 := 0; index26 < size25; index26++ {
            var result27 = buffer.ReadInt()
            var result28 = *buffer.ReadPacket(102).(*ObjectA)
            result24[result27] = result28
        }
    }
    packet.Mm = result24
    var set29 = buffer.ReadIntArray()
    packet.S = set29
    var set30 = buffer.ReadStringArray()
    packet.Ssss = set30
    if buffer.CompatibleRead(beforeReadIndex, length) {
        var result31 = buffer.ReadInt()
        packet.OutCompatibleValue = result31
    }
	if length > 0 {
        buffer.SetReadOffset(beforeReadIndex + length)
    }
	return packet
}

File: protocol/src/test/rust/zfoorust/normalObject.rs
#![allow(dead_code)]
#![allow(unused_imports)]
#![allow(unused_mut)]
#![allow(unused_variables)]
#![allow(non_snake_case)]
#![allow(non_camel_case_types)]
use std::any::Any;
use std::collections::HashMap;
use std::collections::HashSet;
use crate::zfoorust::i_byte_buffer::{IByteBuffer, IPacket};
use crate::zfoorust::objectA::ObjectA;
// 常规的对象，取所有语言语法的交集，基本上所有语言都支持下面的语法
#[derive(Clone)]
pub struct NormalObject {
    pub a: i8,
    pub aaa: Vec<i8>,
    pub b: i16,
    // 整数类型
    pub c: i32,
    pub d: i64,
    pub e: f32,
    pub f: f64,
    pub g: bool,
    pub jj: String,
    pub kk: ObjectA,
    pub l: Vec<i32>,
    pub ll: Vec<i64>,
    pub lll: Vec<ObjectA>,
    pub llll: Vec<String>,
    pub m: HashMap<i32, String>,
    pub mm: HashMap<i32, ObjectA>,
    pub s: HashSet<i32>,
    pub ssss: HashSet<String>,
    pub outCompatibleValue: i32,
    pub outCompatibleValue2: i32,
}

impl IPacket for NormalObject {
    fn protocolId(&self) -> i16 {
        return 101;
    }
}

impl NormalObject {
    pub fn new() -> NormalObject {
        let packet = NormalObject {
            a: 0,
            aaa: Vec::new(),
            b: 0,
            // 整数类型
            c: 0,
            d: 0,
            e: 0f32,
            f: 0f64,
            g: false,
            jj: String::from(""),
            kk: ObjectA::new(),
            l: Vec::new(),
            ll: Vec::new(),
            lll: Vec::new(),
            llll: Vec::new(),
            m: HashMap::new(),
            mm: HashMap::new(),
            s: HashSet::new(),
            ssss: HashSet::new(),
            outCompatibleValue: 0,
            outCompatibleValue2: 0,
        };
        return packet;
    }
}

pub fn writeNormalObject(buffer: &mut dyn IByteBuffer, packet: &dyn Any) {
    let message = packet.downcast_ref::<NormalObject>().unwrap();
    let beforeWriteIndex = buffer.getWriteOffset();
    buffer.writeInt(857);
    buffer.writeByte(message.a);
    buffer.writeByteArray(&message.aaa);
    buffer.writeShort(message.b);
    buffer.writeInt(message.c);
    buffer.writeLong(message.d);
    buffer.writeFloat(message.e);
    buffer.writeDouble(message.f);
    buffer.writeBool(message.g);
    buffer.writeString(message.jj.clone());
    buffer.writePacket(&message.kk, 102);
    buffer.writeIntArray(&message.l);
    buffer.writeLongArray(&message.ll);
    if message.lll.is_empty() {
        buffer.writeInt(0);
    } else {
        buffer.writeInt(message.lll.len() as i32);
        for element0 in message.lll.clone() {
            buffer.writePacket(&element0, 102);
        }
    }
    buffer.writeStringArray(&message.llll);
    buffer.writeIntStringMap(&message.m);
    if message.mm.is_empty() {
        buffer.writeInt(0);
    } else {
        buffer.writeInt(message.mm.len() as i32);
        for (key1, value2) in message.mm.clone() {
            buffer.writeInt(key1);
            buffer.writePacket(&value2, 102);
        }
    }
    buffer.writeIntSet(&message.s);
    buffer.writeStringSet(&message.ssss);
    buffer.writeInt(message.outCompatibleValue);
    buffer.writeInt(message.outCompatibleValue2);
    buffer.adjustPadding(857, beforeWriteIndex);
}

pub fn readNormalObject(buffer: &mut dyn IByteBuffer) -> Box<dyn Any> {
    let length = buffer.readInt();
    let mut packet = NormalObject::new();
    if length == 0 {
        return Box::new(packet);
    }
    let beforeReadIndex = buffer.getReadOffset();
    let result0 = buffer.readByte();
    packet.a = result0;
    let array1 = buffer.readByteArray();
    packet.aaa = array1;
    let result2 = buffer.readShort();
    packet.b = result2;
    let result3 = buffer.readInt();
    packet.c = result3;
    let result4 = buffer.readLong();
    packet.d = result4;
    let result5 = buffer.readFloat();
    packet.e = result5;
    let result6 = buffer.readDouble();
    packet.f = result6;
    let result7 = buffer.readBool(); 
    packet.g = result7;
    let result8 = buffer.readString();
    packet.jj = result8;
    let result9 = buffer.readPacket(102);
    let result10 = result9.downcast_ref::<ObjectA>().unwrap().clone();
    packet.kk = result10;
    let list11 = buffer.readIntArray();
    packet.l = list11;
    let list12 = buffer.readLongArray();
    packet.ll = list12;
    let mut result13: Vec<ObjectA> = Vec::new();
    let size14 = buffer.readInt();
    if size14 > 0 {
        for index15 in 0..size14 {
            let result16 = buffer.readPacket(102);
            let result17 = result16.downcast_ref::<ObjectA>().unwrap().clone();
            result13.push(result17);
        }
    }
    packet.lll = result13;
    let list18 = buffer.readStringArray();
    packet.llll = list18;
    let map19 = buffer.readIntStringMap();
    packet.m = map19;
    let mut result20: HashMap<i32, ObjectA> = HashMap::new();
    let size21 = buffer.readInt();
    if size21 > 0 {
        for index22 in 0..size21 {
            let result23 = buffer.readInt();
            let result24 = buffer.readPacket(102);
            let result25 = result24.downcast_ref::<ObjectA>().unwrap().clone();
            result20.insert(result23, result25);
        }
    }
    packet.mm = result20;
    let set26 = buffer.readIntSet();
    packet.s = set26;
    let set27 = buffer.readStringSet();
    packet.ssss = set27;
    if buffer.compatibleRead(beforeReadIndex, length) {
        let result28 = buffer.readInt();
        packet.outCompatibleValue = result28;
    }
    if buffer.compatibleRead(beforeReadIndex, length) {
        let result29 = buffer.readInt();
        packet.outCompatibleValue2 = result29;
    }
    if length > 0 {
        buffer.setReadOffset(beforeReadIndex + length);
    }
    return Box::new(packet);
}